src/libsyntax/ext/tt/quoted.rs

```diff
@@ -406,71 +457,147 @@ where
             };
 
             if is_1_sep {
-                // #1 is a separator and #2 should be a KleepeOp::*
+                // #1 is a separator and #2 should be a KleepeOp.
                 // (N.B. We need to advance the input iterator.)
                 match parse_kleene_op(input, span) {
-                    // #2 is a KleeneOp (this is the only valid option) :)
-                    Ok(Ok(op)) if op == KleeneOp::ZeroOrOne => {
-                        if !features.macro_at_most_once_rep
-                            && !attr::contains_name(attrs, "allow_internal_unstable")
-                        {
-                            let explain = feature_gate::EXPLAIN_MACRO_AT_MOST_ONCE_REP;
-                            emit_feature_err(
-                                sess,
-                                "macro_at_most_once_rep",
-                                span,
-                                GateIssue::Language,
-                                explain,
-                            );
-                        }
+                    // #2 is `?`, which is not allowed as a Kleene op in 2015 edition.
+                    Ok(Ok((op, op2_span))) if op == KleeneOp::ZeroOrOne => {
+                        sess.span_diagnostic
+                            .struct_span_err(op2_span, "expected `*` or `+`")
+                            .note("`?` is not a macro repetition operator")
+                            .emit();
+
+                        // Return a dummy
+                        return (None, KleeneOp::ZeroOrMore);
+                    }
+
+                    // #2 is a Kleene op, which is the the only valid option
+                    Ok(Ok((op, _))) => {
+                        // Warn that `?` as a separator will be deprecated
+                        sess.span_diagnostic.span_warn(
```

Comment:
Well, it seems like we could issue a lint here -- we just need an `id` -- presumably we can thread the node-id for the macro definition down here? We would then invoke [`buffer_lint`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/session/struct.Session.html#method.buffer_lint), I believe, since HIR lowering has not yet happened:

```rust
sess.buffer_lint(
    LINT_NAME,
    macro_definition_id,
    op1_span,
    "using `?` as a separator is deprecated",
)
```

then we can put the lint into the migration lint set properly.

Reply:
Thanks! I will look into this tomorrow.

Reply:
@nikomatsakis I looked into this a bit more. I am running into two problems:
- `buffer_lint` is not defined on `ParseSess`.
- I don't know where/how to `declare_lint!`.  Doing so requires adding a dependency to `libsyntax` on `librustc`. Is that possible? Assuming it is possible, should I declare the lint in rustc::lint::builtins or in syntax::ext::tt::quoted or somewhere else?